### ISSUE TITLE
Clarification on `function-type-descriptor`

### ISSUE THREAD

**Issue by chiranSachintha** [ballerina-platform/ballerina-spec#870]
**Description:**
Does the following program valid?
(Here, need to know whether default values infer with the type.)

```ballerina
public function main() {
      var foo = function (int a = 5) returns int {
        return a;
    };
    int a = foo();
}
```

**Comment by MaryamZi** [comment]
I believe there's also a similar concern regarding parameter names. Basically, since default values and parameter names aren't part of the shape/type should they be included when inferring a type where `var` is used?

```ballerina
var foo = function (int a = 5) returns int {
    return a;
};
```

Since we infer a type for `var foo`, shouldn't this be similar to defining `foo` with the type-descriptor `function () returns int` rather than `function (int a = 5) returns int` which would allow `foo()` or `foo(a = 1)`?

@jclark, appreciate thoughts on this.

**Comment by jclark** [comment]
I guess the question is whether `var` is inferring a type (a set of shapes) or a type-descriptor (which has stuff like defaults, in addition to a shape).  It's simpler if we say it's inferring a type, so I think we should try that and see if we run into problems.  That would imply that @chiranSachintha's example is not a valid. @MaryamZi WDYT?

**Comment by MaryamZi** [comment]
Yes, that's what I though too (inferring the type instead of the type-descriptor).

But wouldn't this also mean the following should result in a compilation error for missing a required field?

```ballerina
public function main() {
    var x = <Foo> {i: 2};
    x = {}; // `i` is not specified
}

type Foo record {|
    int i = 1;
|};
```

The implementation currently seems to support this; the default value for `i` is used from `Foo`. (The implementation currently seems to be using the type-descriptor).

Even in the function case, I guess there can be scenarios where a user would want to call the same anonymous function with different combinations of arguments for defaultable parameters. If we infer just the type with `var`, an explicit type-descriptor will have to be used when declaring the function variable, to make use of defaultable params in such scenarios (either duplicating the same parameters, default values, etc. and/or having the defaults in the type-descriptor instead of the anonymous function expression) .

IMO, these have an impact on how useful/powerful `var` is. I believe we also currently think of `var` as a direct replacement for specifying the same type-descriptor (as that on the RHS), and if we go with the type approach, this would no longer be true?

**Comment by jclark** [comment]
The overall language direction is *not* to encourage people to use `var` as much as possible.

Your example would be much better written as:

```
Foo x = { i: 2 };
x = {};
```

I don't think inferring a type descriptor is going to be well-defined in conjunction with type narrowing.

```
if x is Foo {

}
else {
   var narrowed = x;
}
```

The type of `x` in the `else` branch will be defined semantically i.e. it is defined in terms of types not type descriptors.

**Comment by MaryamZi** [comment]
> The overall language direction is _not_ to encourage people to use `var` as much as possible.
> 
> Your example would be much better written as:
> 
> ```
> Foo x = { i: 2 };
> x = {};
> ```

Agreed. I think there's common usage where the result of a function call is assigned to a variable defined using `var` (which is a more real-world example compared to what I originally shared), but yes, the same concerns would apply there too. Also in such scenarios, I believe the chances of creating a new value like in the example are quite low.
 
> I don't think inferring a type descriptor is going to be well-defined in conjunction with type narrowing.
> 
> ```
> if x is Foo {
> 
> }
> else {
>    var narrowed = x;
> }
> ```
> 
> The type of `x` in the `else` branch will be defined semantically i.e. it is defined in terms of types not type descriptors.

I agree. So I guess what the implementation is doing right now isn't quite correct. Will create an issue to track fixing this.
